romancal/tweakreg/tweakreg_step.py

```diff
@@ -134,460 +141,331 @@ def process(self, input):
-                bb = image_model.meta.wcs.bounding_box
-                x = catalog["x"]
-                y = catalog["y"]
-                if bb is None:
-                    r, d = image_model.meta.wcs(x, y)
-                    mask = np.isfinite(r) & np.isfinite(d)
-                    catalog = catalog[mask]
-
-                    n_removed_src = np.sum(np.logical_not(mask))
-                    if n_removed_src:
-                        self.log.info(
-                            f"Removed {n_removed_src} sources from {filename}'s "
-                            "catalog whose image coordinates could not be "
-                            "converted to world coordinates."
+                    try:
+                        catalog = self.get_tweakreg_catalog(
+                            source_detection, image_model, i
                         )
-                else:
-                    # assume image coordinates of all sources within a bounding box
-                    # can be converted to world coordinates.
-                    ((xmin, xmax), (ymin, ymax)) = bb
-                    mask = (x > xmin) & (x < xmax) & (y > ymin) & (y < ymax)
-                    catalog = catalog[mask]
-
-                    n_removed_src = np.sum(np.logical_not(mask))
-                    if n_removed_src:
-                        self.log.info(
-                            f"Removed {n_removed_src} sources from {filename}'s "
-                            "catalog that were outside of the bounding box."
-                        )
-
-                # set meta.tweakreg_catalog
-                image_model.meta["tweakreg_catalog"] = catalog.as_array()
-
-                nsources = len(catalog)
-                if nsources == 0:
-                    self.log.warning(f"No sources found in {filename}.")
-                else:
-                    self.log.info(f"Detected {len(catalog)} sources in {filename}.")
-
-                images.shelve(image_model, i)
-
-        # group images by their "group id":
-        group_indices = images.group_indices
-
-        self.log.info("")
-        self.log.info(f"Number of image groups to be aligned: {len(group_indices):d}.")
-        self.log.info("Image groups:")
-
-        imcats = []
-        with images:
-            for i, m in enumerate(images):
-                imcats.append(self._imodel2wcsim(m))
-                images.shelve(m, i, modify=False)
-
-        if len(group_indices) > 1:
-            # local align images:
-            xyxymatch = XYXYMatch(
-                searchrad=self.searchrad,
-                separation=self.separation,
-                use2dhist=self.use2dhist,
-                tolerance=self.tolerance,
-                xoffset=0,
-                yoffset=0,
-            )
+                    except AttributeError as e:
+                        self.log.error(f"Failed to retrieve tweakreg_catalog: {e}")
+                        images.shelve(image_model, i, modify=False)
```

Comment:
If an `AttributeError` exception is raised that means it was not possible to find the catalog in the meta at all. In that case, there's nothing to be modified.

Reply:
Brett, if we leave modify=False here, but do modify the file, what happens?  In this case where we've failed to get the catalog we probably don't want to modify the file.

Mairan, it looks to me like that tweakreg_catalog entry is using an old format we don't want to support (guessing, based on the Table(asarray(...)) wrapping!) so an option would be to just delete this bit if that doesn't have too many other knock-on consequences.

Reply:
> Brett, if we leave modify=False here, but do modify the file, what happens? In this case where we've failed to get the catalog we probably don't want to modify the file.

It depends on if the `ModelLibrary` is `on_disk=True` or `on_disk=False`. For `on_disk=True` passing `modify=False` will prevent writing a temporary file (so if the model was modified a later borrow won't have the modifications). For `on_disk=False` passing `modify=False` does nothing (since the model is already in memory). The general advice is to only use `modify=False` if you're 100% sure there was no modification otherwise it introduces a change in behavior depending on the state of `on_disk`. I agree with @mairanteodoro that the modification doesn't look reachable so `modify=False` is ok here.